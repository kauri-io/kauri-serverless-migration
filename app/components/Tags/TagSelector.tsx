import * as React from 'react'
import TagInput from './TagInput';
import Tag from './Tag';
import { ITag } from './types'
import styled from 'styled-components';

const Container = styled.div`
    display: flex;
    flex-direction: row;
    flex-wrap: wrap;
`;

const Heading = styled.div`
    color: white;
    text-transform: uppercase;
    font-size: ${props => props.theme.fontSizes[0]}px;
    width: 100%;
    font-weight: 500;
    margin-bottom: ${props => props.theme.space[1]}px;
`;

interface IProps {
    maxTags: number;
    availableTags: ITag[];
    tags: string[];
    onChange: (tags: string[]) => void;
    fetchMatches: (val?: string) => void;
}

interface IState {
    maxTags: number;
    currentSelectedTags: string[];
}

class TagSelector extends React.Component<IProps, IState> {
    constructor(props: IProps) {
        super(props);
        this.state = {
            currentSelectedTags: props.tags || [],
            maxTags: props.maxTags,
        }
        this.addTag = this.addTag.bind(this);
        this.removeTag = this.removeTag.bind(this);
        this.handleChange = this.handleChange.bind(this);
        this.handleEnterKey = this.handleEnterKey.bind(this);
        this.removeLastTag = this.removeLastTag.bind(this);
    }

    public componentDidMount() {
        this.props.fetchMatches();
    }

    public addTag(tag: ITag) {
        const prevTags = this.state.currentSelectedTags;
        const currentSelectedTags = prevTags.concat([tag.tag]);
        this.setState({ currentSelectedTags },() => this.props.onChange(currentSelectedTags))
    }

    public removeTag(tag: string) {
        const prevTags = this.state.currentSelectedTags;
        const currentSelectedTags = prevTags.filter(i => i !== tag);
        this.setState({ currentSelectedTags }, () => this.props.onChange(currentSelectedTags))
    }

    public removeLastTag() {
        const tags = this.state.currentSelectedTags;
        tags.pop();
        this.setState({ currentSelectedTags: tags })
    }

    public handleChange(e: React.ChangeEvent<HTMLInputElement>) {
        this.props.fetchMatches(e.target.value);
    }

    public handleEnterKey(val: string) {
        this.addTag({ tag: val.replace(/\s/g, '-'), count: 1})
    }

    public render () {
        return (
            <Container>
                <Heading>Tags Min 1 Max {this.props.maxTags}</Heading>
                {Array.isArray(this.state.currentSelectedTags) && this.state.currentSelectedTags.map(i => <Tag key={i} color="white" removeTag={this.removeTag} tag={i}/>)}
                {this.state.currentSelectedTags.length < this.props.maxTags &&
                    <TagInput
                        selectedTags={this.state.currentSelectedTags}
                        handleEnterKey={this.handleEnterKey}
                        onChange={this.handleChange}
                        onSelect={this.addTag}
                        availableTags={this.props.availableTags}
                        removeLastTag={this.removeLastTag}
                    />
                }
            </Container>
        )
    }
}

export default TagSelector