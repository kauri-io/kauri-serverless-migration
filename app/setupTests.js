import Enzyme, { shallow, mount } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import theme from './lib/theme-config'
import { ThemeProvider } from 'styled-components'

Enzyme.configure({ adapter: new Adapter() })

const observeMock = {
    observe: () => null,
    disconnect: () => null, // maybe not needed
}

window.IntersectionObserver = () => observeMock

export const mountWithTheme = children =>
    mount(<ThemeProvider theme={theme}>{children}</ThemeProvider>)

export const renderWithTheme = children =>
    render(<ThemeProvider theme={theme}>{children}</ThemeProvider>)

export const shallowWithTheme = children =>
    shallow(<ThemeProvider theme={theme}>{children}</ThemeProvider>)
