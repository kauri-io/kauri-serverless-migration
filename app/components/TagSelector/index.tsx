import React from 'react'
import { searchTags } from '../../queries/Tag'
import { Subject, Subscription } from 'rxjs'
import { debounceTime, flatMap, map } from 'rxjs/operators'
import { withApollo } from 'react-apollo'
import { flowRight as compose } from 'lodash'
import { connect } from 'react-redux'
import ApolloClient from 'apollo-client'
import { TagSelector } from '../../components/Tags'
import { ITag } from '../../components/Tags/types'

const handleSearch$: Subject<string> = new Subject()

interface IState {
    tags: string[]
    availableTags: ITag[]
    sub?: Subscription
}

interface IProps {
    client: ApolloClient<{}>
    setFieldsValue: any
    getFieldDecorator: any
    tags: string[]
    updateTags?: (tags: string[]) => void
}

class TagSelectorContainer extends React.Component<IProps, IState> {
    constructor(props: IProps) {
        super(props)
        this.state = {
            availableTags: [],
            tags: props.tags,
        }
        this.updateTags = this.updateTags.bind(this)
    }

    componentDidMount() {
        // Set the field with existing data
        if (this.props.setFieldsValue) {
            this.props.setFieldsValue({ tags: this.state.tags })
        }

        const sub = handleSearch$
            .pipe(
                debounceTime(100),
                flatMap((text: string) =>
                    this.props.client.query<{
                        searchTags: { content: ITag[] }
                    }>({
                        fetchPolicy: 'no-cache',
                        query: searchTags,
                        variables: { query: text, page: 0, size: 10 },
                    })
                ),
                map(
                    ({
                        data: {
                            searchTags: { content },
                        },
                    }) => content
                )
            )
            .subscribe((availableTags: ITag[]) => {
                if (availableTags.length === 0) {
                    availableTags = []
                }
                this.setState({ ...this.state, availableTags })
            })
        this.setState({ ...this.state, sub })
        handleSearch$.next('')
    }

    componentWillUnmount() {
        if (this.state.sub) {
            this.state.sub.unsubscribe()
        }
    }

    updateTags(tags: string[]) {
        this.setState(
            { tags, availableTags: [] },
            () => this.props.updateTags && this.props.updateTags(tags)
        )
    }

    fetchMatches(text?: string) {
        handleSearch$.next(text)
    }

    handleKeyPress(e: React.KeyboardEvent<HTMLInputElement>) {
        e.preventDefault()
    }

    render() {
        return (
            <TagSelector
                tags={this.state.tags}
                fetchMatches={this.fetchMatches}
                onChange={this.updateTags}
                availableTags={this.state.availableTags}
                maxTags={7}
            />
        )
    }
}

const mapStateToProps = () => {
    return {}
}

export default compose(
    withApollo,
    connect(mapStateToProps, {})
)(TagSelectorContainer)
