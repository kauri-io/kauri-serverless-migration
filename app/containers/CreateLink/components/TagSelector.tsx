import React, { CSSProperties, HTMLAttributes } from 'react'
import clsx from 'clsx'
import AsyncSelect from 'react-select/async'
import Typography from '@material-ui/core/Typography'
import NoSsr from '@material-ui/core/NoSsr'
import TextField, { BaseTextFieldProps } from '@material-ui/core/TextField'
import Chip from '@material-ui/core/Chip'
import MenuItem from '@material-ui/core/MenuItem'
import CancelIcon from '@material-ui/icons/Cancel'
import { ValueContainerProps } from 'react-select/src/components/containers'
import { ControlProps } from 'react-select/src/components/Control'
import { NoticeProps } from 'react-select/src/components/Menu'
import { MultiValueProps } from 'react-select/src/components/MultiValue'
import { OptionProps } from 'react-select/src/components/Option'
import { PlaceholderProps } from 'react-select/src/components/Placeholder'
import { Omit } from '@material-ui/types'

import {
    createStyles,
    emphasize,
    makeStyles,
    Theme,
} from '@material-ui/core/styles'
import { searchTags } from '../../../queries/Tag'

export const useStyles = makeStyles((theme: Theme) =>
    createStyles({
        root: {
            margin: theme.spacing(4, 0, 0, 0),
        },
        input: {
            display: 'flex',
            padding: 0,
            height: 'auto',
        },
        valueContainer: {
            display: 'flex',
            flexWrap: 'wrap',
            flex: 1,
            alignItems: 'center',
            overflow: 'hidden',
        },
        chip: {
            margin: theme.spacing(0.5, 0.25),
        },
        chipFocused: {
            backgroundColor: emphasize(
                theme.palette.type === 'light'
                    ? theme.palette.grey[300]
                    : theme.palette.grey[700],
                0.08
            ),
        },
        noOptionsMessage: {
            padding: theme.spacing(1, 2),
        },
        placeholder: {
            position: 'absolute',
            left: 2,
            bottom: 6,
            fontSize: 16,
        },
        label: {
            fontWeight: 600,
        },
    })
)

interface OptionType {
    label: string
    value: string
}

function NoOptionsMessage(props: NoticeProps<OptionType>) {
    return (
        <Typography
            color="textSecondary"
            className={props.selectProps.classes.noOptionsMessage}
            {...props.innerProps}
        >
            {props.children}
        </Typography>
    )
}

type InputComponentProps = Pick<BaseTextFieldProps, 'inputRef'> &
    HTMLAttributes<HTMLDivElement>

function inputComponent({ inputRef, ...props }: InputComponentProps) {
    return <div ref={inputRef} {...props} />
}

function Control(props: ControlProps<OptionType>) {
    const {
        children,
        innerProps,
        innerRef,
        selectProps: { classes, TextFieldProps },
    } = props

    return (
        <TextField
            fullWidth
            InputProps={{
                inputComponent,
                inputProps: {
                    className: classes.input,
                    ref: innerRef,
                    children,
                    ...innerProps,
                },
            }}
            {...TextFieldProps}
        />
    )
}

function Option(props: OptionProps<OptionType>) {
    return (
        <MenuItem
            ref={props.innerRef}
            selected={props.isFocused}
            component="div"
            style={{
                fontWeight: props.isSelected ? 500 : 400,
            }}
            {...props.innerProps}
        >
            {props.children}
        </MenuItem>
    )
}

type MuiPlaceholderProps = Omit<PlaceholderProps<OptionType>, 'innerProps'> &
    Partial<Pick<PlaceholderProps<OptionType>, 'innerProps'>>
function Placeholder(props: MuiPlaceholderProps) {
    const { selectProps, innerProps = {}, children } = props
    return (
        <Typography
            color="textSecondary"
            className={selectProps.classes.placeholder}
            {...innerProps}
        >
            {children}
        </Typography>
    )
}

function ValueContainer(props: ValueContainerProps<OptionType>) {
    return (
        <div className={props.selectProps.classes.valueContainer}>
            {props.children}
        </div>
    )
}

function MultiValue(props: MultiValueProps<OptionType>) {
    return (
        <Chip
            tabIndex={-1}
            label={props.children}
            className={clsx(props.selectProps.classes.chip, {
                [props.selectProps.classes.chipFocused]: props.isFocused,
            })}
            onDelete={props.removeProps.onClick}
            deleteIcon={<CancelIcon {...props.removeProps} />}
        />
    )
}

const components = {
    Control,
    MultiValue,
    NoOptionsMessage,
    Option,
    Placeholder,
    ValueContainer,
}

export default function IntegrationReactSelect({ tags, client, setTags }) {
    const classes = useStyles({})

    const selectStyles = {
        input: (base: CSSProperties) => ({
            ...base,
            '& input': {
                font: 'inherit',
            },
        }),
    }

    return (
        <div className={classes.root}>
            <NoSsr>
                <AsyncSelect
                    classes={classes}
                    styles={selectStyles}
                    inputId="react-select-multiple"
                    TextFieldProps={{
                        label: 'MIX 1 MAX 7 TAGS',
                        InputLabelProps: {
                            htmlFor: 'react-select-multiple',
                            shrink: true,
                            className: classes.label,
                        },
                    }}
                    placeholder="Select up to 7 tags"
                    components={components}
                    getOptionLabel={option => option.label}
                    getOptionValue={option => option.label}
                    value={tags}
                    onChange={val => setTags(val)}
                    loadOptions={inputValue =>
                        new Promise(resolve => {
                            client
                                .query({
                                    fetchPolicy: 'no-cache',
                                    query: searchTags,
                                    variables: {
                                        query: inputValue,
                                        page: 0,
                                        size: 10,
                                    },
                                })
                                .then(res => {
                                    resolve([
                                        ...res.data.searchTags.content.map(
                                            i => ({ label: i.tag })
                                        ),
                                        ...[{ label: inputValue }],
                                    ])
                                })
                        })
                    }
                    isMulti
                />
            </NoSsr>
        </div>
    )
}
