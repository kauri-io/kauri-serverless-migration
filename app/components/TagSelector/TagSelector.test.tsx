import TagSelector from './index'
import { shallowWithTheme } from '../../setupTests'

let props = {}

describe('components/TagSelector', () => {
    it('should match snapshot', () => {
        const wrapper = shallowWithTheme(<TagSelector {...props} />)
        expect(wrapper).toMatchSnapshot()
    })
})
