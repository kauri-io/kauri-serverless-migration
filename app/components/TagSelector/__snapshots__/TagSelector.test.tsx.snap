// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/TagSelector snapshot should match 1`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;
